File: D_Sharky_Surfing.py
import heapq

def min_power_ups_to_finish(n, m, L, hurdles, powerups):
    current_position = 1
    jump_power = 1
    powerup_idx = 0
    powerups_heap = []
    powerup_count = 0
    
    # Step 1: Process hurdles and power-ups
    for i in range(n):
        l, r = hurdles[i]
        
        # While the current position is within the hurdle range, check power-ups
        while current_position < l:
            # Collect all power-ups up to current_position
            while powerup_idx < m and powerups[powerup_idx][0] <= current_position:
                heapq.heappush(powerups_heap, -powerups[powerup_idx][1])  # Store power-ups in max-heap (negate values)
                powerup_idx += 1

            # Try to move forward
            if current_position + jump_power >= l:
                break
            
            # We can't jump far enough, so we need to use power-ups
            if not powerups_heap:
                return -1  # No more power-ups, and we cannot jump far enough
            
            # Collect the largest available power-up
            max_power_up = -heapq.heappop(powerups_heap)  # Get the max power-up
            jump_power += max_power_up
            powerup_count += 1
        
        # Now move to the next hurdle
        if current_position + jump_power < l:
            return -1  # If we can't even make it to the next hurdle
        
        # Update current position
        current_position = r + 1  # Jump after the current hurdle
    
    # Check if we can jump to the end
    while current_position < L:
        while powerup_idx < m and powerups[powerup_idx][0] <= current_position:
            heapq.heappush(powerups_heap, -powerups[powerup_idx][1])
            powerup_idx += 1
        
        if current_position + jump_power >= L:
            break
        
        if not powerups_heap:
            return -1
        
        max_power_up = -heapq.heappop(powerups_heap)
        jump_power += max_power_up
        powerup_count += 1
    
    return powerup_count

def solve():
    t = int(input())
    for _ in range(t):
        n, m, L = map(int, input().split())
        
        hurdles = [tuple(map(int, input().split())) for _ in range(n)]
        powerups = [tuple(map(int, input().split())) for _ in range(m)]
        
        result = min_power_ups_to_finish(n, m, L, hurdles, powerups)
        print(result)

File: D_Sharky_Surfing.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

struct PowerUp {
    int x, v;
};

struct Hurdle {
    int l, r;
};

int minPowerUpsToFinish(int n, int m, int L, vector<Hurdle>& hurdles, vector<PowerUp>& powerups) {
    int currentPosition = 1;
    int jumpPower = 1;
    int powerupCount = 0;
    int powerupIdx = 0;
    
    // Max-heap to store power-ups
    priority_queue<int> powerupsHeap;
    
    // Process hurdles and power-ups
    for (int i = 0; i < n; ++i) {
        int l = hurdles[i].l;
        int r = hurdles[i].r;
        
        // Collect power-ups while current position is less than the hurdle start
        while (currentPosition < l) {
            // Collect all power-ups up to current position
            while (powerupIdx < m && powerups[powerupIdx].x <= currentPosition) {
                powerupsHeap.push(powerups[powerupIdx].v); // Push power-up value
                powerupIdx++;
            }
            
            // Check if we can jump over the hurdle
            if (currentPosition + jumpPower >= l) {
                break;
            }
            
            // If we can't jump over, use the best available power-up
            if (powerupsHeap.empty()) {
                return -1; // No power-ups left and can't jump over
            }
            
            // Collect the largest power-up
            int maxPowerUp = powerupsHeap.top();
            powerupsHeap.pop();
            jumpPower += maxPowerUp;
            powerupCount++;
        }
        
        // Check if we can even make it past the current hurdle
        if (currentPosition + jumpPower < l) {
            return -1; // Can't even make it to the next hurdle
        }
        
        // Move past the current hurdle
        currentPosition = r + 1;
    }
    
    // Check if we can jump to the end L
    while (currentPosition < L) {
        // Collect power-ups up to current position
        while (powerupIdx < m && powerups[powerupIdx].x <= currentPosition) {
            powerupsHeap.push(powerups[powerupIdx].v);
            powerupIdx++;
        }
        
        // Check if we can reach the end L
        if (currentPosition + jumpPower >= L) {
            break;
        }
        
        // Use a power-up if we can't jump far enough
        if (powerupsHeap.empty()) {
            return -1; // Can't proceed and no power-ups left
        }
        
        // Collect the largest power-up
        int maxPowerUp = powerupsHeap.top();
        powerupsHeap.pop();
        jumpPower += maxPowerUp;
        powerupCount++;
    }
    
    return powerupCount;
}

int main() {
    int t;
    cin >> t;
    
    while (t--) {
        int n, m, L;
        cin >> n >> m >> L;
        
        vector<Hurdle> hurdles(n);
        vector<PowerUp> powerups(m);
        
        // Input hurdles
        for (int i = 0; i < n; ++i) {
            cin >> hurdles[i].l >> hurdles[i].r;
        }
        
        // Input power-ups
        for (int i = 0; i < m; ++i) {
            cin >> powerups[i].x >> powerups[i].v;
        }
        
        int result = minPowerUpsToFinish(n, m, L, hurdles, powerups);
        cout << result << endl;
    }
    
    return 0;
}
